resources/type-system/inference.md

```diff
@@ -345,15 +349,22 @@ all `return` and `yield` statements in the block body have been considered.
 Let `T` be the **actual returned type** of a function literal as computed above.
 Let `R` be the greatest closure of the typing context `K` as computed above.
 
+*Now compute `S`, which is the future value type of an `async` function, the
+element type of a generator function, and the return type of other functions.*
+
 With null safety: if `R` is `void`, or the function literal is marked `async`
 and `R` is `FutureOr<void>`, let `S` be `void` (without null-safety: no special
 treatment is applicable to `void`).
 
-Otherwise, if `T <: R` then let `S` be `T`.  Otherwise, let `S` be `R`.  The
-inferred return type of the function literal is then defined as follows:
+If the previous paragraph did not yield a value for `S`: When the function is
+marked `async`: if `T <: futureValueType(R)` then let `S` be `T`; otherwise let
+`S` be `futureValueType(R)`. When the function is not marked `async`: if `T <:
+R` then let `S` be `T`; otherwise let `S` be `R`.
```

Comment:
We need to use `futureValueType(R)`, not `flatten(R)`. If we were to use flatten then we'd carry `?` over from the context. So we could have a context type that wants a `Future<num>?`, then we'd get `int?` as the value of `R`, and then we'd get an inferred return type of `Future<int?>` where the context wants a return type of `Future<int>?`, and that's a type error.

```dart
Future<int>? Function() f = () async => 1;
```

In this example `R` is `Future<int>?` and `flatten(R)` is `int?`. But we don't want the function literal to have a return type of `Future<int?>`, it should be `Future<int>`.